lib/tbot/config/service_kubernetes_v2.go

```diff
@@ -0,0 +1,161 @@
+	_ ServiceConfig = &KubernetesV2Output{}
+	_ Initable      = &KubernetesV2Output{}
+)
+
+const KubernetesV2OutputType = "kubernetes/v2"
+
+// KubernetesOutput produces credentials which can be used to connect to a
+// Kubernetes Cluster through teleport.
+type KubernetesV2Output struct {
+	// Destination is where the credentials should be written to.
+	Destination bot.Destination `yaml:"destination"`
+
+	// Roles is the list of roles to request for the generated credentials.
+	// If empty, it defaults to all the bot's roles.
+	Roles []string `yaml:"roles,omitempty"`
+
+	// DisableExecPlugin disables the default behavior of using `tbot` as a
+	// `kubectl` credentials exec plugin. This is useful in environments where
+	// `tbot` may not exist on the system that will consume the outputted
+	// kubeconfig. It does mean that kubectl will not be able to automatically
+	// refresh the credentials within an individual invocation.
+	DisableExecPlugin bool `yaml:"disable_exec_plugin,omitempty"`
+
+	// Selectors is a list of selectors for path-based routing. Multiple
+	// selectors can be used to generate an output containing all matches.
+	Selectors []*KubernetesSelector `yaml:"selectors,omitempty"`
+}
+
+func (o *KubernetesV2Output) CheckAndSetDefaults() error {
+	if err := validateOutputDestination(o.Destination); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(o.Selectors) == 0 {
+		return trace.BadParameter("at least one selector must be provided")
+	}
+
+	for _, s := range o.Selectors {
+		if err := s.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return trace.Wrap(o.Destination.CheckAndSetDefaults())
+}
+
+func (o *KubernetesV2Output) GetDestination() bot.Destination {
+	return o.Destination
+}
+
+func (o *KubernetesV2Output) Init(ctx context.Context) error {
+	return trace.Wrap(o.Destination.Init(ctx, []string{}))
+}
+
+func (o *KubernetesV2Output) Describe() []FileDescription {
+	// Based on tbot.KubernetesOutputService.Render
+	return []FileDescription{
+		{
+			Name: "kubeconfig.yaml",
+		},
+		{
+			Name: IdentityFilePath,
+		},
+		{
+			Name: HostCAPath,
+		},
+	}
+}
+
+func (o *KubernetesV2Output) MarshalYAML() (interface{}, error) {
+	type raw KubernetesV2Output
+	return withTypeHeader((*raw)(o), KubernetesV2OutputType)
+}
+
+func (o *KubernetesV2Output) UnmarshalYAML(node *yaml.Node) error {
+	dest, err := extractOutputDestination(node)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Alias type to remove UnmarshalYAML to avoid recursion
+	type raw KubernetesV2Output
+	if err := node.Decode((*raw)(o)); err != nil {
+		return trace.Wrap(err)
+	}
+	o.Destination = dest
+	return nil
+}
+
+func (o *KubernetesV2Output) Type() string {
+	return KubernetesV2OutputType
+}
+
+// KubernetesSelector allows querying for a Kubernetes cluster to include either
+// by its name or labels.
+type KubernetesSelector struct {
+	Name string `yaml:"name,omitempty"`
+
+	Labels map[string]string `yaml:"labels,omitempty"`
```

Comment:
When specifying selectors throughout the rest of Teleport, we usually seem to support multiple values being provided as an OR. Perhaps we ought to support that here for consistency.

Reply:
There's always a chance I'm just missing a better API, but we don't seem to expose any generic RPC for querying resources with a list of label values from a client. `ListResources` only supports a `map[string]string` label selector, sadly. This API also doesn't accept wildcards, which is unfortunate.

That does put this at odds with #50812 which has explicit server-side support for full label queries through its RPC. What do you think, does this limitation pose a significant UX problem?

Reply:
Ah - in which case, let's leave it as is for now. I'd figured that the ListResources API would support this but apparently not. It should be possible to migrate from `map[string]string` to `map[string][]string` at a later date anyway.